src/accel/FastSimulationOffload.cc

```diff
@@ -0,0 +1,88 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file accel/FastSimulationOffload.cc
+//---------------------------------------------------------------------------//
+#include "FastSimulationOffload.hh"
+
+#include "corecel/Assert.hh"
+
+#include "ExceptionConverter.hh"
+#include "LocalTransporter.hh"
+#include "SharedParams.hh"
+
+namespace celeritas
+{
+FastSimulationOffload::FastSimulationOffload(G4String const& name,
+                                             SharedParams const* params,
+                                             LocalTransporter* local)
+    : G4VFastSimulationModel(name), params_(params), transport_(local)
+{
+    CELER_VALIDATE(G4VERSION_NUMBER > 1110,
+                   << "the current version of Geant4 (" << G4VERSION_NUMBER
+                   << ") is too old to support the fast simulation Flush() "
+                      "interface");
+    CELER_EXPECT(params);
+    CELER_EXPECT(local);
+}
+
+FastSimulationOffload::FastSimulationOffload(G4String const& name,
+                                             G4Envelope* region,
+                                             SharedParams const* params,
+                                             LocalTransporter* local)
+    : G4VFastSimulationModel(name, region), params_(params), transport_(local)
+{
+    CELER_VALIDATE(G4VERSION_NUMBER > 1110,
+                   << "the current version of Geant4 (" << G4VERSION_NUMBER
+                   << ") is too old to support the fast simulation Flush() "
+                      "interface");
+    CELER_EXPECT(region);
+    CELER_EXPECT(params);
+    CELER_EXPECT(local);
+}
+
+G4bool FastSimulationOffload::IsApplicable(G4ParticleDefinition const& particle)
+{
+    CELER_EXPECT(*params_);
+ 
+    return (std::find(params_->OffloadParticles().begin(),
+                      params_->OffloadParticles().end(),
+                      &particle)
+            != params_->OffloadParticles().end());
+}
+
+G4bool FastSimulationOffload::ModelTrigger(G4FastTrack const& /*track*/)
+{
+    // At present we don't offload based on kinematics (or anything else
+    // available via G4FastTrack), so always return true
+    return true;
+}
+
+void FastSimulationOffload::DoIt(G4FastTrack const& track, G4FastStep& step)
+{
+    // Offload this track to Celeritas for transport
+    CELER_EXPECT(track.GetPrimaryTrack());
+    CELER_EXPECT(*transport_);
+ 
+    ExceptionConverter call_g4exception{"celer0001", params_};
```

Comment:
Should we reuse the same "code" as in the `./SimpleOffload` (since presumably we're not going to have both simultaneously); or do we want to make the codes unique over all `accel` classes?

Reply:
Hmm, I guess the problem with unique codes over all of `accel` would be tracking them and adding new ones with the correct incremented number? Or could we just require each code to be unique within a class, with some magic to turn `celer0001` into, say `accel_FastSimulationOffload_celer0001`.

I don't have a strong opinion either way...

Reply:
If they're unique within a class then either the codes need unique prefixes or the same code needs to have the same meaning if it's used multiple times. Ultimately these aren't very important because they just show up in the user's G4Exception code, and even Geant4 doesn't consistently use codes. (Some have a prefix, some are debugging information, etc.) We can take a minute to discuss at the meeting but I think they would have the most utility if they're unique across the codebase: presumably the most common use case is a user saying "I encountered error celer1234"? Even then the associated exception information is the important part...